### ISSUE TITLE
Support for per rule exclusions in azurerm_web_application_firewall_policy

### ISSUE THREAD

**Issue by sirsnork** [hashicorp/terraform-provider-azurerm#16934]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

When using the OWASAP 3.2 ruleset in a WAF you're able to create per rule exclusions as outlined [here](https://docs.microsoft.com/en-us/azure/web-application-firewall/ag/application-gateway-waf-configuration?tabs=portal#per-rule-exclusions).

It would be really useful to be able to do this in TF. Obviously this would only be useable on OWASP 3.2 and above

Thanks

### New or Affected Resource(s)/Data Source(s)

azurerm_web_application_firewall_policy

### Potential Terraform Configuration

```hcl
resource "azurerm_web_application_firewall_policy" "example" {
...
    managed_rules {
        managed_rule_set {
            type    = "OWASP"
            version = "3.2"
            rule_exclusion {
                rule_ids = [
                    "920300",
                ]
                match_variable          = "RequestCookieNames"
                selector                = "too-tasty"
                selector_match_operator = "EndsWith"
            }
        }
    }
}
```

**Comment by sinbai** [comment]
@sirsnork  thank you for opening this issue. Per [the latest terraform provider documentation](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/web_application_firewall_policy#managed_rules), it seems that terraform provider supports the above feature . Could you follow the example in the documentation to see if it helps?

**Comment by hs-rittmeier** [comment]
Hello, we could use the same feature.

@sinbai As far as I can say the 3.7 azurerm doesnt have any support for adding rule_ids to the exclusion
![image](https://user-images.githubusercontent.com/89528589/170248670-0d186c0f-eaee-413a-b726-9b1dc773ccfb.png)

**Comment by sirsnork** [comment]
As hs-rittmeier says, there doesn't appear to be any way to list the rule id's even with a custom exclusion, nor any indication that even if there was it would only work with OWASP 3.2 and later so either the documentation is incomplete or that does not support the functionality

**Comment by sinbai** [comment]
Could you try using `disabled_rules ` to specify exclusion rule ids like below?

```
 managed_rule_set {
      type    = "OWASP"
      version = "3.2"
      rule_group_override {
        rule_group_name = "REQUEST-920-PROTOCOL-ENFORCEMENT"
        disabled_rules = [
          "920300"
        ]
      }
    }
```
>

**Comment by sirsnork** [comment]
I already do, but that disables the entire rule. Opening a much larger hole. Larger even than using a global exception potentially

Per rule exclusions allow you exclude a single variable/value in the query string or JSON payload from being processed by the rule(s), allowing you to leave the rule on for all other traffic.

**Comment by sinbai** [comment]
@sirsnork yes, you are right. Currently, terraform does not support [`exclusionManagedRuleSets`](https://github.com/Azure/azure-rest-api-specs/blob/ef5e61721cf3ce245ec78f2908da1aced5a83b42/specification/network/resource-manager/Microsoft.Network/stable/2021-08-01/webapplicationfirewall.json#L804). That's what you need, right?

**Comment by sirsnork** [comment]
@sinbai honestly I'm not sure if that is it, I don't use ARM for anything so decoding that I couldn't be sure. 

From the original link it's this that I'd like to see

> Per-rule exclusions
> You can configure an exclusion for a specific rule, group of rules, or rule set. You must specify the rule or rules that the exclusion applies to. You also need to specify the request attribute that should be excluded from the WAF evaluation.
> 
> Per-rule exclusions are available when you use the OWASP (CRS) ruleset version 3.2 or later.

**Comment by kamil-makowski** [comment]
Are there any plans to implement this feature in near future?

**Comment by JustinGalbraith** [comment]
Is this feature on the roadmap to add to AzureRM? My company had a long discussion this week about the pros and cons of disabling rules vs this sort of targeted exclusion. Having this available in TF would ease a lot of people concerns about the current tuning process (per-path rule disablement). 

Thanks

**Comment by 0x4c6565** [comment]
> Is this feature on the roadmap to add to AzureRM? My company had a long discussion this week about the pros and cons of disabling rules vs this sort of targeted exclusion. Having this available in TF would ease a lot of people concerns about the current tuning process (per-path rule disablement).
> 
> Thanks

I'm going to look at adding this as we have a requirement for this also

**Comment by sinbai** [comment]
@sirsnork It seems that Terraform has been supported the request. This is the [Terraform doc](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/web_application_firewall_policy#excluded_rule_set). Could you follow up to see if it works?

**Comment by 0x4c6565** [comment]
I've just started to look at implementing this, however I can see this is already supported. I've used the below config as an example:

```hcl
resource "azurerm_resource_group" "testrg" {
  name     = "testrg"
  location = "UK West"
}

resource "azurerm_web_application_firewall_policy" "testwafpolicy" {
  name                = "testwafpolicy"
  resource_group_name = azurerm_resource_group.testrg.name
  location            = azurerm_resource_group.testrg.location

  policy_settings {
    enabled                     = false
    file_upload_limit_in_mb     = 100
    max_request_body_size_in_kb = 128
    mode                        = "Detection"
    request_body_check          = true
  }

  managed_rules {
    managed_rule_set {
      type    = "OWASP"
      version = "3.2"
    }

    exclusion {
      match_variable          = "RequestHeaderNames"
      selector_match_operator = "Equals"
      selector                = "x-some-header"

      excluded_rule_set {
        type    = "OWASP"
        version = "3.2"
        
        rule_group {
          rule_group_name = "REQUEST-942-APPLICATION-ATTACK-SQLI"
          excluded_rules = [
            942430,
            942440
          ]
        }
      }
    }
  }
}
```

Confirmed exclusions added:

![image](https://user-images.githubusercontent.com/7476213/218782127-ae8ac838-de07-4059-94d9-743b7f196a94.png)

**Comment by sinbai** [comment]
@0x4c6565 thanks for your feedback. @sirsnork given this feature has been supported by Terraform, are you okey if we close this issue?

**Comment by rcskosir** [comment]
Thanks for taking the time to submit this issue. It looks like this has been resolved as of #17757. As such, I am going to mark this issue as closed. If that is not the case, please provide additional information including the version in which you are still experiencing this issue, thanks!
